Fixes #12156 - Adds email configurations into Foreman settings

amirfefer opened the pull request:
Instead using email.yaml, the email configurations can be modified dynamically

amirfefer committed:
Fixes #12156 - Adds email configurations into Foreman settings

Instead using email.yaml, the email configurations
 can be modified dynamically via settings

ohadlevy commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,19 @@
+    self.transaction do
+      [
+        self.set('email_yaml', N_("Keeps email.yaml configurations"), false, N_('YAML conifguration')),
+        self.set('delivery_method', N_("Delivery method can be smtp or sendmail"), 'smtp', N_('Delivery Method')),
```
is there a way to migrate existing settings?

amirfefer replied:
@ohadlevy  Do you mean something like  `db:seed`?

amirfefer replied:
Existing email settings will migrate into default values in the settings table.

domcleal commented:
Marking as WoC due to above comment and test failures.

amirfefer commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,19 @@
+        self.set('smtp_port', N_("SMTP server port"), smtp_options[:port], N_('Port')),
+        self.set('smtp_auth', N_("Authentication type, can be none, plain, login and cram_md5."), smtp_options[:authentication], N_('Authentication')),
+        self.set('smtp_user', N_("Required when the authentication type is not none"), smtp_options[:user_name], N_('Username')),
+        self.set('smtp_password', N_("Required when the authentication type is not none"), smtp_options[:password], N_('Password'))
```
Is it necessary to encrypt this field?

domcleal replied:
Yes, probably.

domcleal commented on app/mailers/application_mailer.rb:
```diff
@@ -1,7 +1,9 @@
 class ApplicationMailer < ActionMailer::Base
+  after_filter :set_configurations unless Setting[:email_yaml]
   default :from => Proc.new { Setting[:email_reply_address] || "noreply@foreman.example.org" }
+  self.delivery_method = Setting[:delivery_method] unless Setting[:email_yaml]
```
These checks on the email_yaml setting are being made at the class level, so will it require a restart to take effect?  Can they be dynamic?

domcleal commented on app/mailers/application_mailer.rb:
```diff
@@ -11,6 +13,16 @@ def mail(headers = {}, &block)
 
+  def smtp_options
+    {
+      address:              Setting[:smtp_address],
```
excess whitespace, should use `=>` key/value separator for consistency

domcleal commented on app/mailers/application_mailer.rb:
```diff
@@ -28,4 +40,8 @@ def set_url
     end
   end
+
+  def set_configurations
```
Poorly named, try `configure_delivery_method_from_settings` or similar.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,19 @@
+    method = ActionMailer::Base.delivery_method
+    self.transaction do
+      [
+        self.set('email_yaml', N_("Uses email.yaml configurations instead"), false, N_('YAML conifguration')),
```
1) typo in the setting name
2) "configurations" reads poorly
3) "instead" of what?

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,19 @@
+    self.transaction do
+      [
+        self.set('email_yaml', N_("Uses email.yaml configurations instead"), false, N_('YAML conifguration')),
+        self.set('delivery_method', N_("Delivery method can be smtp or sendmail"), method, N_('Delivery Method')),
```
"Method" should be lower case for consistency

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,19 @@
+      [
+        self.set('email_yaml', N_("Uses email.yaml configurations instead"), false, N_('YAML conifguration')),
+        self.set('delivery_method', N_("Delivery method can be smtp or sendmail"), method, N_('Delivery Method')),
+        self.set('smtp_address', N_("SMTP server address"), smtp_options[:address], N_('Address')),
```
"Address" is too vague, it's an SMTP server hostname or similar.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,19 @@
+        self.set('delivery_method', N_("Delivery method can be smtp or sendmail"), method, N_('Delivery Method')),
+        self.set('smtp_address', N_("SMTP server address"), smtp_options[:address], N_('Address')),
+        self.set('smtp_port', N_("SMTP server port"), smtp_options[:port], N_('Port')),
+        self.set('smtp_auth', N_("Authentication type, can be none, plain, login and cram_md5."), smtp_options[:authentication], N_('Authentication')),
```
Authentication method, surely?  Not just "Authentication"

domcleal commented:
1) This is missing various advanced SMTP and sendmail settings, is email.yaml meant to be used instead?

2) Lots of settings are missing from test fixtures.

3) Very little test coverage, it's not testing all of the settings and doesn't test the setting's load_defaults method.

4) All of the settings with limited values should be validated.

5) It looks like an upgrade will simply ignore email.yaml immediately, shouldn't the default value of email_yaml depend on whether it's already configured?

stbenjam commented:
@amirfefer Hi, this has been sitting around for a while, so I was wondering if you planned to update it? I'd really like to see this feature added, as the email.yaml is such a pain for users to update.

I was fiddling around with it and I have some code suggestions here: https://github.com/theforeman/foreman/compare/develop...stbenjam:email

You're welcome to take them into this PR, or I can open a new one.

My branch adds onto your changes these things:
- Settings are imported from email.yaml if it exists, and they are marked read-only
- Adds support for sendmail options, too
- Adds more SMTP settings
- Adds test coverage

amirfefer commented:
@stbenjam Thanks! I'll look into this in the next few days

amirfefer commented:
@domcleal Could you review this PR?

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,72 @@
+  def self.delivery_settings
+    options = {}
+    self.all.each do |setting|
+      if Setting[:delivery_method] == :smtp && setting.name =~ /^smtp_/
```
Prefer `String#start_with?` instead of using a regexp.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,72 @@
+    self.all.each do |setting|
+      if Setting[:delivery_method] == :smtp && setting.name =~ /^smtp_/
+        options[setting.name.to_s.gsub(/^smtp_/, '')] = setting.value
+      elsif Setting[:delivery_method] == :sendmail && setting.name =~ /^sendmail_/
```
ditto, `.start_with?`

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,72 @@
+    File.expand_path(File.join(File.dirname(__FILE__), '..', '..', '..', 'config', 'email.yaml'))
+  end
+
+  def self.delivery_settings
```
Missing its own unit test coverage, only seems to be tested a little from the mailer? The whole return value can be tested directly (for both delivery methods too), I think.

domcleal replied:
Coverage for :smtp seems to be there now, but still missing coverage of the :sendmail branch.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,72 @@
+        self.set('email_reply_address', N_("Email reply address for emails that Foreman is sending"), email_reply_address, N_('Email reply address')),
+        self.set('email_subject_prefix', N_("Prefix to add to all outgoing email"), '[foreman]', N_('Email subject prefix')),
+        self.set('send_welcome_email', N_("Send a welcome email including username and URL to new users"), false, N_('Send welcome email')),
+        self.set('delivery_method', N_("Method used to deliver e-mail"), 'sendmail', N_('Delivery method'), nil, { :collection => Proc.new {{'sendmail' => :sendmail, 'smtp' => :smtp}}}),
```
a) "e-mail" has no hyphen in Foreman
b) "smtp" should be "SMTP", it's an acronym

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,72 @@
+        self.set('smtp_tls', N_("Enable SSL/TLS"), false, N_('SMTP use SSL/TLS')),
+        self.set('smtp_address', N_("Address to connect to"), '', N_('SMTP address')),
+        self.set('smtp_port', N_("Port to connect to"), 25, N_('SMTP port')),
+        self.set('smtp_domain', N_("Email domain"), '', N_('SMTP email domain')),
```
The description is ambiguous, this is the EHLO domain, right? Not the from domain?

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,72 @@
+        self.set('smtp_domain', N_("Email domain"), '', N_('SMTP email domain')),
+        self.set('smtp_user_name', N_("Username to use to authenticate, if required"), '', N_('SMTP username')),
+        self.set('smtp_password', N_("Password to use to authenticate, if required"), '', N_('SMTP password'), nil, {:encrypted => true}),
+        self.set('smtp_authentication', N_("Specify authentication type here, if required"), 'none', N_('SMTP authentication'), nil, { :collection => Proc.new {{'plain' => :plain, 'login' => :login, 'cram_md5' => :cram_md5, 'none' => :none}}}),
```
"here" is superfluous

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,72 @@
+  end
+
+  def self.lookup_value(name)
+    if name =~ /^smtp_/ && mailconfig[:smtp_settings].is_a?(Hash)
```
prefer .start_with?, ditto below

domcleal commented on app/views/settings/index.html.erb:
```diff
@@ -10,6 +11,11 @@
 <div class="tab-content">
   <% @settings.group_by(&:category).each do |category, setting| %>
     <div class="tab-pane <%= "active" if category == @settings.first.category %>" id='<%= short_cat(category) %>' >
+      <% if category == "Setting::Email" %>
```
Adding a partial per setting type would be much better (see also, compute resources pages).

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,72 @@
+require 'facter'
+
+class Setting::Email < Setting
+  WHITELIST = %w(send_welcome_email email_reply_address email_subject_prefix)
```
`NON_EMAILYAML_SETTINGS` or similar? Whitelist isn't descriptive.

domcleal commented on test/mailers/application_mailer_test.rb:
```diff
@@ -1,7 +1,13 @@
 class ApplicationMailerTest < ActiveSupport::TestCase
-  setup { ActionMailer::Base.deliveries = [] }
+  setup do
+    setup do
```
Too many setup blocks?

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,72 @@
+        self.set('email_subject_prefix', N_("Prefix to add to all outgoing email"), '[foreman]', N_('Email subject prefix')),
+        self.set('send_welcome_email', N_("Send a welcome email including username and URL to new users"), false, N_('Send welcome email')),
+        self.set('delivery_method', N_("Method used to deliver e-mail"), 'sendmail', N_('Delivery method'), nil, { :collection => Proc.new {{'sendmail' => :sendmail, 'smtp' => :smtp}}}),
+        self.set('smtp_enable_starttls_auto', N_("Use STARTTLS automatically"), false, N_('SMTP enable StartTLS auto')),
```
"SMTP automatic STARTTLS" may be a better name.

domcleal commented:
Thanks for the update, that's quite a lot of change, mostly good. I'm concerned that there's not enough test coverage of the modifications you've made to readonly attributes and loading email.yaml in the Setting::Email subclass.

Can you ensure the Setting::Email methods have full test coverage? It's easy to mock the loading of YAML files, then test the parsing methods all work as expected.

domcleal reviewed (changes requested):
Some inline comments, but unit test coverage needs to be much higher for the new Setting::Email functionality.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,72 @@
+  end
+
+  def has_readonly_value?
+    self.class.mailconfig.present? && WHITELIST.exclude?(name)
```
Should this only be read only when the setting is also set in email.yaml?

amirfefer commented:
@domcleal I fixed the requests changes above.

domcleal commented on test/models/setting_test.rb:
```diff
@@ -561,4 +560,41 @@ def check_value_returns_from_cache_with(options = {})
     refute_valid bmc, :base, 'Unable to disable safemode_render when bmc_credentials_accessible is disabled'
   end
+
+  context "when email.yaml exists" do
```
These should be in a test case specifically for Setting::Email, probably in `test/models/setting/`?

domcleal commented on test/models/setting_test.rb:
```diff
@@ -180,7 +180,6 @@ def test_attributes_in_SETTINGS_are_readonly
   def test_value_is_updated_after_change_in_SETTINGS
     setting_name = "foo_#{rand(1000000)}"
     Setting.create!(:name => setting_name, :value => "bar", :default => "default", :description => "foo")
-
```
No need to change this.

domcleal commented on config/email.yaml.example:
```diff
@@ -1,4 +0,0 @@
-# Outgoing email settings
```
Why remove this? If you're deprecating it then it makes some sense to, but it's still fully supported from what I understand. (Do you mean to deprecate it?)

domcleal commented on app/views/settings/index.html.erb:
```diff
@@ -2,6 +2,7 @@
 <% title_actions documentation_button('3.5.2ConfigurationOptions') %>
 <ul class="nav nav-tabs" data-tabs="tabs">
   <% @settings.group_by(&:category).each do |category, setting| %>
+
```
Unnecessary?

domcleal commented on app/assets/javascripts/settings.js:
```diff
@@ -8,3 +8,26 @@ $(document).ready(function() {
+function test_mail(item, email, url) {
+  $(item).addClass("disabled");
+  $('#test_indicator').show();
+  var email = email
```
Unnecessary?

domcleal commented on app/assets/javascripts/settings.js:
```diff
@@ -8,3 +8,26 @@ $(document).ready(function() {
+      $('#test_indicator').hide();
+      $(item).removeClass("disabled");
+    }
+  })
```
end with ;

domcleal reviewed (changes requested):
Please also rebase, there are some conflicts.

amirfefer commented:
@domcleal  Thanks for the review, I've rebased, and fix the requests above. I've improved test coverage for `sendmail` in `email_settings_test` (I've separated email-related tests in `settings_test` to `test/model/settings` dir),  plus I've added `sendmail_location` setting and  `email.yaml` configuration has been deprecated.

domcleal commented on app/helpers/settings_helper.rb:
```diff
@@ -33,4 +33,8 @@ def translate_full_name(setting)
     trunc_with_tooltip(fullname, 32, setting.name, false)
   end
+
+  def render_if_partial_exists(path, setting)
```
This is quite similar to `provider_partial_exist?`, and given the generic name I think it should move into one of the core layout helpers, or be settings-specific and have a settings-specific name prefix.

domcleal commented on app/helpers/settings_helper.rb:
```diff
@@ -33,4 +33,8 @@ def translate_full_name(setting)
   end
+
+  def render_if_partial_exists(path, setting)
+    render path, :setting => setting if lookup_context.find_all(path,[],true).any?
```
`lookup_context.exists?` might be shorter. Missing spaces between arguments.

domcleal commented on test/models/settings/email_setting_test.rb:
```diff
@@ -0,0 +1,53 @@
+
+class EmailSettingTest < ActiveSupport::TestCase
+  def setup
+    Setting.cache.clear
```
This is already in the `reset_setting_cache` setup/teardown method, is it needed?

domcleal commented on test/models/settings/email_setting_test.rb:
```diff
@@ -0,0 +1,53 @@
+    test 'should use smtp settings' do
+      Setting::Email.create(:name => :delivery_method, :value => :smtp, :category => "Setting::Email")
+      Setting::Email.create(:name => :smtp_address, :value => "sm@example.com", :category => "Setting::Email")
+      Setting::Email.create(:name => :smtp_authentication, :value => "none", :category => "Setting::Email")
```
Are these lines doing anything? Should they instead be `Setting[:smtp_authentication] = 'none'` etc?

I'm confused because the assertion on the line below checks if the delivery_settings equal the ones defined in the hash above, but not that it contains any of these modifications you're attempting to make on lines 21-23.

amirfefer replied:
@domcleal  The modification I've made in lines 21-23 should simulate `load_defualt` method, and  those settings are read only (using yaml config), so the values should update only by `mailconfig` (the yaml file).
I've change this test to be more clear.

domcleal commented on test/models/settings/email_setting_test.rb:
```diff
@@ -0,0 +1,53 @@
+
+    test 'should use sendmail settings' do
+      @delivery_settings["delivery_method"] = :sendmail
+      Setting::Email.create(:name => :delivery_method, :value => :sendmail, :category => "Setting::Email")
```
ditto, isn't this already defined in fixtures?

domcleal commented on app/views/settings/index.html.erb:
```diff
@@ -10,6 +10,7 @@
 <div class="tab-content">
   <% @settings.group_by(&:category).each do |category, setting| %>
     <div class="tab-pane <%= "active" if category == @settings.first.category %>" id='<%= short_cat(category) %>' >
+      <%= render_if_partial_exists "settings/category/#{category.sub(/.*?::/, '').downcase}", setting %>
```
`.demodulize` instead of sub and regexp

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+  NON_EMAIL_YAML_SETTINGS = %w(send_welcome_email email_reply_address email_subject_prefix)
+
+  def self.load_defaults
+    Foreman::Deprecation.deprecation_warning('1.16', "email configuration has been moved to settings") if mailconfig.present?
```
1. 1.15 at this stage (last stable branch + 2, allows it to be released in two stable releases)
2. Please improve the deprecation warning, specifying that it's email.yaml that has been deprecated.

domcleal commented on app/views/settings/category/_email.html.erb:
```diff
@@ -0,0 +1,3 @@
+<%= link_to_function(_("Test email"), "test_mail(this, '#{User.current.mail}', '#{test_mail_user_url(User.current)}')",
```
Single quotes are permitted in email addresses, the `User.current.mail` should be escaped or use the ID.

amirfefer commented:
Thanks @domcleal for your review, I've fixed those issues above, please have a look at
the inline comment in `email_setting_test`.

I've changed the `test_mail` in `users_controller` a bit, so it  supports on getting user id in params too.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+        self.set('smtp_password', N_("Password to use to authenticate, if required"), '', N_('SMTP password'), nil, {:encrypted => true}),
+        self.set('smtp_authentication', N_("Specify authentication type, if required"), 'none', N_('SMTP authentication'), nil, { :collection => Proc.new {{'plain' => :plain, 'login' => :login, 'cram_md5' => :cram_md5, 'none' => :none}}}),
+        self.set('sendmail_arguments', N_("Specify additional options to sendmail"), '-i', N_('Sendmail arguments')),
+        self.set('sendmail_location', N_("The location of the sendmail executable"), '', N_('Sendmail location'))
```
defaults to `/usr/sbin/sendmail`, without it, the command isn't set by default (unlike with email.yaml)

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+        self.set('email_reply_address', N_("Email reply address for emails that Foreman is sending"), email_reply_address, N_('Email reply address')),
+        self.set('email_subject_prefix', N_("Prefix to add to all outgoing email"), '[foreman]', N_('Email subject prefix')),
+        self.set('send_welcome_email', N_("Send a welcome email including username and URL to new users"), false, N_('Send welcome email')),
+        self.set('delivery_method', N_("Method used to deliver email"), 'sendmail', N_('Delivery method'), nil, { :collection => Proc.new {{'Sendmail' => :sendmail, 'SMTP' => :smtp}}}),
```
1. Shouldn't the keys and values be swapped around? I see the symbols in the UI, not the nicer names. Values from email.yaml aren't being shown correctly because of it.
2. Extract the Sendmail/SMTP strings with `N_()`.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+        self.set('smtp_domain', N_("HELO/EHLO domain"), '', N_('SMTP HELO/EHLO domain')),
+        self.set('smtp_user_name', N_("Username to use to authenticate, if required"), '', N_('SMTP username')),
+        self.set('smtp_password', N_("Password to use to authenticate, if required"), '', N_('SMTP password'), nil, {:encrypted => true}),
+        self.set('smtp_authentication', N_("Specify authentication type, if required"), 'none', N_('SMTP authentication'), nil, { :collection => Proc.new {{'plain' => :plain, 'login' => :login, 'cram_md5' => :cram_md5, 'none' => :none}}}),
```
1. Swap keys and values again.
2. `none` isn't valid (`ArgumentError: wrong authentication type none`), it should be removed from both the possible values and the default, then instead let the user clear the value.
3. Even when cleared, I think the authentication type can't be sent as an empty string - it has to be left out of the configuration, else you get a similar error for an empty string.

amirfefer replied:
@domcleal, Now authentication type (and other email settings) can be cleared  and left out the configuration.

domcleal commented on app/models/setting.rb:
```diff
@@ -10,7 +10,7 @@ class Setting < ActiveRecord::Base
   FROZEN_ATTRS = %w{ name category full_name }
   NONZERO_ATTRS = %w{ puppet_interval idle_timeout entries_per_page max_trend outofsync_interval }
-  BLANK_ATTRS = %w{ trusted_puppetmaster_hosts login_delegation_logout_url authorize_login_delegation_auth_source_user_autocreate root_pass default_location default_organization websockets_ssl_key websockets_ssl_cert oauth_consumer_key oauth_consumer_secret login_text}
+  BLANK_ATTRS = %w{ sendmail_location trusted_puppetmaster_hosts login_delegation_logout_url authorize_login_delegation_auth_source_user_autocreate root_pass default_location default_organization websockets_ssl_key websockets_ssl_cert oauth_consumer_key oauth_consumer_secret login_text smtp_address smtp_domain smtp_user_name smtp_password smtp_authentication sendmail_arguments}
```
Nitpick, put sendmail_location with the other sendmail_args items

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+  def self.delivery_settings
+    options = {}
+    self.all.each do |setting|
+      if Setting[:delivery_method] == :smtp && setting.name.start_with?('smtp')
```
`Setting[:delivery_method]` may be a string if set through the UI, but may be a symbol if set through email.yaml. If you change it in the UI, these if statements don't match and `options = {}`.

amirfefer replied:
`Setting[:delivery_method]` value is a symbol now if set through the UI.

domcleal commented on app/mailers/application_mailer.rb:
```diff
@@ -2,7 +2,9 @@
 class ApplicationMailer < ActionMailer::Base
   include Roadie::Rails::Automatic
-  default :from => Proc.new { Setting[:email_reply_address] || "noreply@foreman.example.org" }
+  default :raise_delivery_errors => true, :delivery_method => Proc.new { Setting[:delivery_method] },
```
I don't think `raise_delivery_errors` can be set here, it just seems to create a header:

```
raise-delivery-errors: true
```

shouldn't it be in the environment config?

domcleal commented on app/assets/javascripts/settings.js:
```diff
@@ -8,3 +8,25 @@ $(document).ready(function() {
+    },
+    error: function (xhr) {
+      var error = $.parseJSON(xhr.responseText).message;
+      notify("<p>" + error + "</p>", 'error');
```
This doesn't seem to render correctly:

![screenshot from 2016-11-11 15-17-10](https://cloud.githubusercontent.com/assets/482089/20219580/05f441b2-a822-11e6-9e70-000502daea69.png)

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+  NON_EMAIL_YAML_SETTINGS = %w(send_welcome_email email_reply_address email_subject_prefix)
+
+  def self.load_defaults
+    Foreman::Deprecation.deprecation_warning('1.15', "'email.yaml' file configuration has been deprecated, email configurations has been moved to foreman settings") if mailconfig.present?
```
1. Capitalise 'Foreman'
2. "configurations" => "configuration"

domcleal reviewed (changes requested):
I'm finding quite a few small bugs when changing the various delivery options, they need a bit more testing.

amirfefer commented:
@domcleal  Thanks for the review.
I've fixed the issues above, and I've added a test for checking empty email settings value.

domcleal commented:
> Setting[:delivery_method] value is a symbol now if set through the UI.

I don't think this is the case, it's still set as a string when updating through the UI. Can you re-test this? Every time I set it and query the value, it's a string. It shouldn't be possible to set a symbol through the form + controllers in any case.

Setting::Email.delivery_settings still does symbol-only comparisons, which I think needs to handle both.

domcleal commented:
> Setting::Email.delivery_settings still does symbol-only comparisons, which I think needs to handle both.

Please remember to update tests for this if you change it.

amirfefer commented:
@domcleal  `Setting::Email.delivery_settings` is  handling both string and symbol situations now.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+  def self.delivery_settings
+    options = {}
+    self.all.each do |setting|
+      if Setting[:delivery_method].to_s == "smtp" && setting.name.start_with?('smtp') && setting.value.present?
```
Using `setting.value.present?` is omitting settings that are `false` - e.g. set enable StartTLS auto to false and it isn't returned in the `Setting::Email.delivery_settings` return value so the setting has no effect. (Because it defaults to `true` in Rails.)

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+        self.set('email_subject_prefix', N_("Prefix to add to all outgoing email"), '[foreman]', N_('Email subject prefix')),
+        self.set('send_welcome_email', N_("Send a welcome email including username and URL to new users"), false, N_('Send welcome email')),
+        self.set('delivery_method', N_("Method used to deliver email"), 'sendmail', N_('Delivery method'), nil, { :collection => Proc.new {{:sendmail => N_("Sendmail"), :smtp => N_("SMTP")}}}),
+        self.set('smtp_enable_starttls_auto', N_("SMTP automatic STARTTLS"), false, N_('SMTP enable StartTLS auto')),
```
enable_starttls_auto defaults to `true` in Rails

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+        self.set('send_welcome_email', N_("Send a welcome email including username and URL to new users"), false, N_('Send welcome email')),
+        self.set('delivery_method', N_("Method used to deliver email"), 'sendmail', N_('Delivery method'), nil, { :collection => Proc.new {{:sendmail => N_("Sendmail"), :smtp => N_("SMTP")}}}),
+        self.set('smtp_enable_starttls_auto', N_("SMTP automatic STARTTLS"), false, N_('SMTP enable StartTLS auto')),
+        self.set('smtp_tls', N_("Enable SSL/TLS"), false, N_('SMTP use SSL/TLS')),
```
What uses this setting?

amirfefer replied:
Changed to  `openssl_verify_mode`.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+        self.set('smtp_domain', N_("HELO/EHLO domain"), '', N_('SMTP HELO/EHLO domain')),
+        self.set('smtp_user_name', N_("Username to use to authenticate, if required"), '', N_('SMTP username')),
+        self.set('smtp_password', N_("Password to use to authenticate, if required"), '', N_('SMTP password'), nil, {:encrypted => true}),
+        self.set('smtp_authentication', N_("Specify authentication type, if required"), 'none', N_('SMTP authentication'), nil, { :collection => Proc.new {{:plain => 'plain', :login => 'login', :cram_md5 => 'cram_md5', '' => N_("none")}}}),
```
Using `N_("none")` here in the collection hash (and for delivery_method) is notionally correct, but the content doesn't seem to be translated anywhere. Enable mark_translated and the text is not marked.

amirfefer replied:
I've used `_()` instead of `N_` because the hash is a proc, and I assume it execute only when it renders to the user.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,74 @@
+      if Setting[:delivery_method].to_s == "smtp" && setting.name.start_with?('smtp') && setting.value.present?
+        options[setting.name.to_s.gsub(/^smtp_/, '')] = setting.value
+      elsif Setting[:delivery_method].to_s == "sendmail" && setting.name.start_with?('sendmail') && setting.value.present?
+        options[setting.name.to_s.gsub(/^sendmail_/, '')] = setting.value
```
Note that CodeClimate is stating this method is complex, I'd suggest deduplicating it if possible. Ditto with `lookup_value`.

domcleal reviewed (changes requested):
Disabling "Enable StartTLS auto" isn't working correctly, i18n only partly implemented. Please check CodeClimate results and try to improve the introduced issues (complex load_defaults is OK).

amirfefer commented:
@domcleal  I've fixed the changes requested.
Please look at the inline comments

domcleal commented on config/email`.yaml:
```diff
@@ -0,0 +1,21 @@
+# Outgoing email settings
```
You should delete this file and change your password!

amirfefer replied:
Oops, I shouldn't update a PR at 4 AM. Thanks for your attention

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,81 @@
+        self.set('send_welcome_email', N_("Send a welcome email including username and URL to new users"), false, N_('Send welcome email')),
+        self.set('delivery_method', N_("Method used to deliver email"), 'sendmail', N_('Delivery method'), nil, { :collection => Proc.new {{:sendmail => _("Sendmail"), :smtp => _("SMTP")}}}),
+        self.set('smtp_enable_starttls_auto', N_("SMTP automatic STARTTLS"), true, N_('SMTP enable StartTLS auto')),
+        self.set('smtp_openssl_verify_mode', N_("When using TLS, you can set how OpenSSL checks the certificate"), '', N_('SMTP OpenSSL verify mode'), nil, { :collection => Proc.new {{:none => _("none"), :peer => _("peer"), :client_once => _("client_once"), :fail_if_no_peer_cert => _("fail_if_no_peer_cert")}}}),
```
1. What should the default value of this be? `none` is different to the default value of `''`, is an empty default option meant to be a selectable value? If so, what is the default?
1. Nitpick, here you're translating the human-readable names of the constants but in smtp_authentication similar constants are untranslated. Either make them English strings that can easily be translated or don't translate the constants.

amirfefer replied:
Yes, the empty default should be selectable, otherwise this setting always get into `delivery_settings'.

domcleal reviewed (changes requested):
email.yaml file should not be committed, some questions about the new openssl_verify_mode setting.

amirfefer commented:
@domcleal  Thanks for your review, I've added a selectable default to `smtp_openssl_verify_mode`.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,81 @@
+        self.set('send_welcome_email', N_("Send a welcome email including username and URL to new users"), false, N_('Send welcome email')),
+        self.set('delivery_method', N_("Method used to deliver email"), 'sendmail', N_('Delivery method'), nil, { :collection => Proc.new {{:sendmail => _("Sendmail"), :smtp => _("SMTP")}}}),
+        self.set('smtp_enable_starttls_auto', N_("SMTP automatic STARTTLS"), true, N_('SMTP enable StartTLS auto')),
+        self.set('smtp_openssl_verify_mode', N_("When using TLS, you can set how OpenSSL checks the certificate"), '', N_('SMTP OpenSSL verify mode'), nil, { :collection => Proc.new {{'' => _("no OpenSSL certificate verifier"), :none => _("none"), :peer => _("peer"), :client_once => _("client_once"), :fail_if_no_peer_cert => _("fail_if_no_peer_cert")}}}),
```
1. "no OpenSSL certificate verifier" should probably be "Default verification mode", which looks like "peer". Verification isn't disabled by default. I would probably keep it as a separate value to peer though, that's set inside OpenSSL.
1. "fail_if_no_peer_cert" etc are not good English strings to translate, either use proper descriptions or don't translate them.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,81 @@
+        self.set('smtp_domain', N_("HELO/EHLO domain"), '', N_('SMTP HELO/EHLO domain')),
+        self.set('smtp_user_name', N_("Username to use to authenticate, if required"), '', N_('SMTP username')),
+        self.set('smtp_password', N_("Password to use to authenticate, if required"), '', N_('SMTP password'), nil, {:encrypted => true}),
+        self.set('smtp_authentication', N_("Specify authentication type, if required"), 'none', N_('SMTP authentication'), nil, { :collection => Proc.new {{:plain => _("plain"), :login => _("login"), :cram_md5 => _("cram_md5"), '' => _("none")}}}),
```
"plain", "login" etc are SMTP methods that will not make sense translated as individual words out of context, either make this clearer in the strings ("PLAIN method", and possibly add a translator's comment above) or don't translate.

domcleal reviewed (changes requested):
Unsure that the translated constants make sense still, please consider what they should be in both English and another language.

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,81 @@
+  NON_EMAIL_YAML_SETTINGS = %w(send_welcome_email email_reply_address email_subject_prefix)
+
+  def self.load_defaults
+    Foreman::Deprecation.deprecation_warning('1.15', "'email.yaml' file configuration has been deprecated, email configuration has been moved to Foreman settings") if mailconfig.present?
```
Sorry, I think I stated earlier that this should be 1.15, but it should be 1.16 at the time of writing. (Last stable [1.13] plus three: https://theforeman.org/handbook.html#Do)

If it's merged into develop now, it's then deprecated in 1.14, 1.15 and then removed in 1.16. If it's merged after 1.14's branching then it would be set to 1.17.

amirfefer commented:
@domcleal  I agree it better to keep those parameters with no translation. It can be confusing otherwise

domcleal commented on app/models/setting/email.rb:
```diff
@@ -0,0 +1,81 @@
+        self.set('smtp_domain', N_("HELO/EHLO domain"), '', N_('SMTP HELO/EHLO domain')),
+        self.set('smtp_user_name', N_("Username to use to authenticate, if required"), '', N_('SMTP username')),
+        self.set('smtp_password', N_("Password to use to authenticate, if required"), '', N_('SMTP password'), nil, {:encrypted => true}),
+        self.set('smtp_authentication', N_("Specify authentication type, if required"), 'none', N_('SMTP authentication'), nil, { :collection => Proc.new {{:plain => "plain", :login => "login", :cram_md5 => "cram_md5", '' => _("none")}}}),
```
The default of `'none'` here should probably be `''` (none is the caption)?

amirfefer commented:
@domcleal , I've rebased and updated the default value in `smtp_authentication`.

domcleal commented:
@amirfefer Sorry, could you rebase again please? The commit it's based on had an error (fixed in #17494) causing tests not to run properly, it needs updating.

amirfefer commented:
@domcleal , I've rebased, could you re-review please?

domcleal commented:
Merged, thanks @amirfefer.
